### ISSUE TITLE
Gemma 4 vision encoder: performance improvement

### ISSUE THREAD

**Issue by gkumbhat** [torch-spyre/spyre-inference#976]
## Summary

The Gemma 4 vision tower costs **the same per image regardless of the image's resolution**.
A 192x192 image yields 16 usable soft tokens and pays what a 768x768 one paying for 256
does.

Three separable causes, measured below. The largest one is coupled to a compile-shape
constraint, so the fixes have a required ordering.

## Measurements

27 layers, real `google/gemma-4-26B-A4B-it` vision config, random weights (the tower is
stock `transformers` code either way, so this measures the same graph a 26B checkpoint
builds). Per-stage, with a device sync either side of each stage. Reported relative to the
cheapest case, since only the ratios matter here.

| patches | grid | soft tokens | total, relative |
|---|---|---|---|
| 2520 (as shipped) | 48x48 | 256 | 3.7x |
| 2520 (as shipped) | 24x24 | 64 | 3.7x |
| 576 (trimmed) | 24x24 | 64 | 1.07x |
| 2520 (as shipped) | 12x12 | 16 | 3.7x |
| 144 (trimmed) | 12x12 | 16 | 1.0x (baseline) |

Rows 2-vs-3 and 4-vs-5 are the same image with and without the padding. The three padded
rows are indistinguishable from each other: resolution does not affect cost at all today.

Where the time goes: at the full pad the encoder is ~94% of the forward and the patch
embedder ~5%; once trimmed to a small grid the encoder is ~82% and the patch embedder ~17%.
Pooler and tail are under 1% in every case -- don't spend effort there.

## Finding 1 - every image is padded to the maximum patch count (3.7x waste)

`transformers/models/gemma4/image_processing_gemma4.py:220` computes
`max_patches = max_soft_tokens * pooling_kernel_size**2` = `280 * 9` = **2520**, and line
261 pads *every* image to it with `(-1, -1)` position rows. `Gemma4VisionModel.forward`
then derives `output_length` from the padded length, so all 27 layers run at 2520 patches
(2560 after the stick pad) whatever the real grid. vLLM's `gemma4_mm.py:135` documents the
same, and `MultiModalFieldConfig.batched("image")` wants uniform shapes.

The pad is roughly 70% of wall-clock for anything below 768x768.

Trimming is compatible with the tower's own constraints: the encoder only needs a uniform
mask, and the pooler only needs `9 * output_length == num_patches`, so any multiple of 9
works. Multiples of 576 (LCM of 9 and 64, so the stick pad stays free) give a 5-rung
ladder: 576 / 1152 / 1728 / 2304 / 2520.

## Finding 2 - a length-independent floor from eager per-op dispatch

Encoder cost is affine in patch count with a large intercept: extrapolated to zero patches
it still costs about a quarter of its per-layer cost at the full 2520 pad, and 144 patches
and 576 patches pay that floor identically.

That floor is per-op dispatch. The tower runs eager
(`compilation_config.compile_mm_encoder` is `False`), so torch-spyre lowers each aten op
as its own kernel, and `_run_layer` in `spyre_inference/multimodal/gemma4_vision.py` is
30+ ops -- several of them avoidable:

- `padded_sdpa` does 3x `F.pad` of q/k/v per layer purely for the 2520->2560 alignment,
  i.e. 81 full-tensor copies per image.
- `_apply_rope` adds two `[L,128] x [128,128]` permutation matmuls per layer.
- `_padded_attn_mask` materialises a `[1,1,2560,2560]` fp16 mask per image where O(L) would
  do.
- `_pool_weights` carries an explicit `TODO: rebuilt per image`.

After Finding 1 is fixed this floor becomes the large majority (~82%) of what remains, so
the two compose rather than compete.

## Finding 3 - a full position-table device->host copy per forward, for a trivial gather

`spyre_inference/multimodal/gemma4_vision.py:675` (`_position_embeddings`) runs

```python
table = convert(self.position_embedding_table, device="cpu")
```

on **every** forward. The table is `[2, position_embedding_size=10240, hidden_size=1152]`,
i.e. **47.2 MB** in fp16, and at most 2520 of its rows are ever gathered (~96 distinct for
a 48x48 grid).

Measured in isolation on card, that copy is **two orders of magnitude** more expensive than
the host-side gather it exists to feed, and it accounts for essentially the whole
patch-embed stage (see the stage shares above). The table also pins 47.2 MB of device
memory although its only consumer is a host-side gather.

Fix: place it on the host permanently at patch time, exactly as `place_vision_tail_on_cpu`
already does for the tail. Few lines, no numerical change.

## The constraint: the tower is outside both bucketers, and compiles lazily in the serving path

Every distinct tower *device shape* is a fresh lazy compile, taken on the request that
first presents it. A new shape costs well over an order of magnitude more than a
steady-state forward -- tens of seconds of stall, in the serving path. Observed in the same
sweep's first-iteration timings:

```
2520 patches, first ever seen      : compile (cold Inductor caches, the most expensive)
2520 patches, different valid grid : cache hit, indistinguishable from steady state
 576 patches, first seen           : compile
2520 patches again                 : cache hit
 144 patches, first seen           : compile
```

Positions and masks are host-side, so varying *contents* costs nothing -- only device
shapes matter.

Why nothing catches this:

- `SpyreAttnBucketer` (KV/query buckets) only covers vLLM `Attention` layers with a KV
  cache. The tower uses `padded_sdpa` on a raw HF `Gemma4VisionAttention`; no vLLM
  `Attention` layer is ever built. `grep -rn "bucket" spyre_inference/multimodal/` is
  empty.
- `spyre_shape_bucketer` covers the decoder body.
  `TorchSpyreModelRunner.warming_up_model()` walks those buckets, the logits row widths,
  and `_record_attention_graphs()` -- nothing calls the tower.
- vLLM's `profile_run`, which *does* feed dummy multimodal data, never runs:
  `SpyreWorker.determine_available_memory` returns the fixed KV budget and short-circuits
  it.

So the first real image request pays that compile.

### Today the patch axis is accidentally safe; the batch axis is not

Because the processor pads everything to 2520 there is exactly one patch-count shape: one
compile, then never again. **The waste in Finding 1 is the flip side of that shape
stability**, which is what makes it more than a one-line fix.

The image-count axis has no such protection. `gemma4_mm.py::_process_image_input` calls the
encoder with `pv_tensor` of shape `[chunk, patches, pp]` where
`chunk = min(len(items), _encoder_chunk(...))`. On this card the free/total device memory
that `_encoder_chunk` reads puts its cap in the hundreds of images, so the cap is never hit
in practice: **chunk is just however many images arrive in that scheduler step** -- up to
`max_num_seqs x limit_mm_per_prompt`, and which value appears depends on scheduling. Every
distinct N is a distinct device shape and another compile stall.

(Aside: `_encoder_chunk` sizes that cap from the `F.one_hot` transient in
`_position_embeddings`, which our `patch_patch_embedder` already eliminates -- so the
formula is budgeting for memory we do not spend.)

Also: `tests/e2e/test_multimodal.py`'s gemma-4 case covers one image only, so batch > 1
through the patched tower is untested for correctness, never mind compile cost.

## Proposed ordering

| # | change | effect | risk |
|---|---|---|---|
| 1 | Host-resident `position_embedding_table` (Finding 3) | removes the patch-embed stage almost entirely, frees 47.2 MB of device memory | trivial, local, no shape implications |
| 2 | Warm the tower in `warming_up_model()` for the shapes already reachable (batch 1..K at 2520) | moves the first-image compile out of the serving path; removes the per-N stall | needs a dummy tower run in the runner |
| 3 | Bucket patch counts to the 576 ladder + warm the cross-product (Finding 1) | ~3.5x on sub-768x768 images | touches the processor output; only pays off *after* 2 |
| 4 | Prune per-layer ops / compile the encoder block (Finding 2) | attacks the dispatch floor, which dominates once 3 lands | compile-time and lowering risk |

3 must not land before 2: on its own it trades a per-image win for a tens-of-seconds stall
on every new resolution a client sends. vLLM's own `buckets.setdefault(pv.shape[0], ...)`
grouping in `_process_image_input` is already built for multiple patch counts, so the
engine side needs nothing new -- the work is our processor emitting bucketed lengths plus
the warmup pass.

## Reproducing

Measured with `scripts/bench_gemma4_vision.py` (not yet landed; on the
`gemma4-multimodal-vision` branch). It builds `Gemma4VisionModel` from the real vision
config with random weights and applies
`spyre_inference.multimodal.gemma4_vision.apply`, so no 26B checkpoint or engine boot is
needed, then times `patch_embed` / `encoder` / `pooler` / `tail` separately:

```
uv run --no-sync python scripts/bench_gemma4_vision.py --sweep --iters 3
```

`--patches 2520 --grid 12x12` is what a small image costs today; `--patches 144 --grid
12x12` is what it would cost trimmed.
